### PR TITLE
Switch to OSRRelease() throughout code

### LINKED ISSUE #441: GPKG writing crasher
## Expected behavior and actual behavior.

Segmentation faults often occur when writing a small GeoPackage file.

```bash
$ python gpkg.py
Warning 6: dataset foo.gpkg does not support layer creation option ENCODING
Segmentation fault: 11
```

I don't immediately see a connection between the warning and the segfaut because the warning always occurs and the segfaults do not always occur.

The problem was reported by conda/conda-forge user @chryss at https://github.com/conda-forge/fiona-feedstock/issues/41 and I've been able to trigger it using the wheels on PyPI.

## Steps to reproduce the problem.

gpkg.py below:

```python
import fiona
schema = {'geometry': 'Point', 'properties': [('title', 'str')]}     
crs = {'init': 'EPSG:4326'}    
with fiona.open('foo.gpkg', 'w', layer='bar_foo', driver='GPKG',     
schema=schema, crs=crs) as d:    
    d.writerecords([{    
                    'geometry': {'type': 'Point', 'coordinates': [0.0, 0.0]},    
                    'properties': {'title': 'One'}}])    
    d.writerecords([{    
                    'geometry': {'type': 'Point', 'coordinates': [2.0, 3.0]},    
                    'properties': {'title': 'Two'}}])
```

## Operating system

* Mac OS X 10.11.6 and Python 2.7 (@chryss)
* Mac OS X 10.12.3 and Python 3.6.0 (@sgillies)

## Fiona version and provenance

1.7.5 wheel from PyPI: macosx and py36.

### PR DESCRIPTION
**Pull request by sgillies**
Also: use a `ctypedef void * OGRSpatialReferenceH`.

Resolves #441.

### CODE REVIEW

File: fiona/ogrext1.pyx [hunk 1]
```diff
@@ -771,9 +771,9 @@ cdef class WritingSession(Session):
 
                 elif collection.name is None:
                     ogrext1.OGR_DS_Destroy(cogr_ds)
-                    cogr_ds == NULL
+                    cogr_ds = NULL
```
Comment (sgillies) [inline fiona/ogrext1.pyx]: Found this when taking a closer look at compiler warning!

File: fiona/ogrext1.pyx [hunk 2]
```diff
@@ -451,7 +451,7 @@ cdef class Session:
 
     def stop(self):
         self.cogr_layer = NULL
-        if self.cogr_ds is not NULL:
+        if self.cogr_ds != NULL:
```
Comment (sgillies) [inline fiona/ogrext1.pyx]: More reliable.

File: fiona/ogrext2.pyx [hunk 3]
```diff
@@ -417,7 +418,7 @@ cdef class Session:
             flags = ogrext2.GDAL_OF_VECTOR | ogrext2.GDAL_OF_READONLY
             try:
                 self.cogr_ds = ogrext2.GDALOpenEx(
-                    path_c, flags, drvs, NULL, NULL)
+                    path_c, flags, <const char *const *>drvs, NULL, NULL)
```
Comment (sgillies) [inline fiona/ogrext2.pyx]: To suppress a compiler warning.

**Comment by sgillies** [comment]
@snorfalorpagus possible for you to review this for me?

**Comment by snorfalorpagus** [comment]
@sgillies Yes I can review it tomorrow.

File: fiona/ogrext2.pyx [hunk 4]
```diff
@@ -877,11 +880,14 @@ cdef class WritingSession(Session):
                 collection.driver == "ESRI Shapefile" and
                 'ISO-8859-1') or sysencoding).upper()
 
+            # The ENCODING option makes no sense for some drivers and
+            # will result in a warning. Fixing is a TODO.
```
Comment (snorfalorpagus) [inline fiona/ogrext2.pyx]: Need an issue for this.

**Comment by sgillies** [comment]
I got tripped up by warnings about `GDAL_DATA` and `PROJ_LIB` in a couple test results due to some logging I added, but explicitly exporting these variables in `.travis.yml` solved that.

**Pull request merged**